third_party/move/tools/move-linter/src/model_ast_lints/nonminimal_bool.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements an expression linter that checks for boolean identities
+//! that occur in conditionals:
+//! 1. `x && true`, which can be replaced with just `x`.
+//! 2. `x && false`, which can be replaced with just `false`.
+//! 3. `x || true`, which can be replaced with just `true`.
+//! 4. `x || false`, which can be replaced with just `x`.
+//! 5. `x <==> true`, which can be replaced with just `x`.
+//! 6. `x <==> false`, which can be replaced with just `!x`.
+//! 7. `x ==> true`, which can be replaced with just `true`.
+//! 8. `x ==> false`, which can be replaced with just `!x`.
+//! 9. `true ==> x`, which can be replaced with just `x`.
+//! 10. `false ==> x`, which can be replaced with just `true`.
+//! 11. `!true`, which can be replaced with just `false`.
+//! 12. `!false`, which can be replaced with just `true`.
+//!
+//! Note also that rules 1 through 6 have both LHS and RHS version
+
+use move_compiler_v2::external_checks::ExpChecker;
+use move_model::{
+    ast::{ExpData, Operation, Value},
+    model::GlobalEnv,
+};
+
+#[derive(Default)]
+pub struct NonMinimalBool;
+
+impl ExpChecker for NonMinimalBool {
+    fn get_name(&self) -> String {
+        "nonminimal_bool".to_string()
+    }
+
+    fn visit_expr_pre(&mut self, env: &GlobalEnv, expr: &ExpData) {
+        use ExpData::{Call, Value as ExpValue};
+        use Operation::*;
+        use Value::Bool;
+
+        let get_msg = |cmp: &Operation, b: bool, side| {
+            match (cmp, b, side) {
+            (And, true, s) | (Or, false, s) => Some(format!(
+                "The {} occurrence of `{}` is redundant and should be removed.",
+                if s == 0 { "left" } else { "right" },
+                b
+            )),
+            (And, false, _) | (Or, true, _) | (Implies, true, 1) => {
+                Some(format!("This expression can be simplified as `{}`.", b))
```

Comment:
Consider the example code:
```
    fun mutate(x: &mut u64): bool {
        *x = 1;
        true
    }

    fun test_mut() {
        let x = 0;
        mutate(&mut x) && false;
    }
```

For this code, we incorrectly suggest:
```
+ warning: [lint] This expression can be simplified as `false`.
+    ┌─ tests/model_ast_lints/nonminimal_bool.move:57:9
+    │
+    │         mutate(&mut x) && false;
+    │         ^^^^^^^^^^^^^^^^^^^^^^^
+    │
```

Simplifying to `false` changes the semantics of the code. While one could do additional checks for purity, perhaps the easier thing in this PR is to suggest something like "This expression always evaluates to `false` and can likely be simplified".

Reply:
I see. I suppose there's also a question of if we have something like ``if(1/0 && false)``, aborting is different than ``false``. So ``if(unknown_function() && false)`` really only "partially" evaluates to false.

Reply:
That's true (although it is less likely to have real-world code like `1/0 && false`, compared the the test). My main concern (in the test case I provided) is that some one applies the linter and blindly follows the explicit suggestion, then it could cause semantic changes.